examples/bls12-381/src/bls12-381.rs

```diff
@@ -0,0 +1,794 @@
+    let t0 = fp2inv(t0); //t0^-1
+    let x = fp2mul(x0, t0); //x0 * t0
+    let y = fp2mul(y0, t0); // y0 * t0
+    let z = fp2mul(z0, t0); // z0 * t0
+    (x, y, z)
+    
+
+}
+
+/* Arithmetic for Fp12 elements */
+// Algorithms from: https://eprint.iacr.org/2010/354.pdf
+pub fn fp12fromfp6(n: Fp6) -> Fp12 {
+    (n, fp6zero())
+}
+
+
+pub fn fp12neg(n: Fp12) -> Fp12 {
+    let (n1, n2) = n;
+    (fp6sub(fp6zero(), n1), fp6sub(fp6zero(), n2))
+}
+
+pub fn fp12add(n: Fp12, m: Fp12) -> Fp12 {
+    let (n1, n2) = n;
+    let (m1, m2) = m;
+    (fp6add(n1, m1), fp6add(n2, m2))
+}
+
+pub fn fp12sub(n: Fp12, m: Fp12) -> Fp12 {
+    fp12add(n, fp12neg(m))
+}
+
+pub fn fp12mul(n: Fp12, m: Fp12) -> Fp12 {
+    let (n1, n2) = n;
+    let (m1, m2) = m;
+    let gamma = (fp2zero(), fp2fromfp(Fp::ONE()), fp2zero()); //0 + v + 0 (c0, c1v, c2v^2)
+    
+    let t1 = fp6mul(n1, m1); //n1 * n2
+    let t2 = fp6mul(n2, m2); //n2 * m2
+    let x = fp6add(t1, fp6mul(t2, gamma)); //t1 + (t2 * gamma)
+    let y = fp6mul(fp6add(n1, n2), fp6add(m1, m2)); //(n1 + n2) * (m1 + m2)
+    let y = fp6sub(fp6sub(y, t1), t2); //y - t1 - t2
+    (x, y)
+}
+
+pub fn fp12inv(n: Fp12) -> Fp12 {
+    let (n1, n2) = n;
+    let gamma = (fp2zero(), fp2fromfp(Fp::ONE()), fp2zero()); //0 + v + 0 (c0, c1v, c2v^2)
+    
+    let t1 = fp6mul(n1, n1); //n1^2
+    let t2 = fp6mul(n2, n2); //n2^2
+    let t1 = fp6sub(t1, fp6mul(gamma, t2)); //t1 - (gamma * t2)
+    let t2 = fp6inv(t1); //t1^-1
+    let x = fp6mul(n1, t2); //n1 * t2
+    let y = fp6neg(fp6mul(n2, t2)); //-(n2 * t2)
+    (x, y)
+}
+
+pub fn fp12exp(n: Fp12, k: Scalar) -> Fp12 {
+    let l = 255 - most_significant_bit(k, 255);
+    let mut c = n;
+    for i in l..255 { //starting from second most significant bit
+        c = fp12mul(c, c);
+        if k.bit(255-i-1) {
+            c = fp12mul(c, n);
+        }
+    }
+    c
+}
+
+pub fn fp12conjugate(n: Fp12) -> Fp12 {
+    let (n1, n2) = n;
+    (n1, fp6neg(n2))
+}
+
+pub fn fp12zero() -> Fp12 {
+    fp12fromfp6(fp6zero())
+}
+
+/* Arithmetic in G1 */
+
+//g1 add with no Point at Infinity
+fn g1add_a(p: G1, q: G1) -> G1
+{
+    let (x1, y1, _) = p;
+    let (x2, y2, _) = q;
+
+    let x_diff = x2 - x1; 
+    let y_diff = y2 - y1;
+    let xovery = y_diff * x_diff.inv(); //  x / y = x * y^-1
+    let x3 = xovery.exp(2u32) - x1 - x2;
+    let y3 = xovery * (x1 - x3) - y1;
+    (x3, y3, false)
+}
+
+//g1 double with no Point at Infinity
+fn g1double_a(p: G1) -> G1
+{
+    let (x1, y1, _) = p;
+    
+    let x12 = x1.exp(2u32);
+    let xovery = (Fp::from_literal(3u128) * x12) * (Fp::TWO() * y1).inv();
+    let x3 = xovery.exp(2u32) - Fp::TWO()*x1;
+    let y3 = xovery * (x1 - x3) - y1;
+    (x3, y3, false)
+}
+/* Wrapper functions with Point of Infinity */
+pub fn g1add(p: G1, q: G1) -> G1 {
+    let (x1, y1, inf1) = p;
+    let (x2, y2, inf2) = q;
+
+    if inf1 {
+        q
+    } else { if inf2 {
+        p
+    } else { if p == q {
```

Comment:
Hum so `Eq/Neq` is tricky because it is polymorphic and we have to make sure that the equality semantics in Rust is the same as the polymorphic structural equality semantic matches in F* or EC. For instance it two `Seq` `x` aqnd `y` have the same contents but are not allocated at the same time, will we have `x == y` in Rust ?

Reply:
Well, it's currently allowed by the typechecker but can't be translated to F*/EC. So we have to make this consistent in some way.

> For instance it two `Seq` `x` aqnd `y` have the same contents but are not allocated at the same time, will we have `x == y` in Rust ?

Rust `==` will only look at content. If we want it to do anything on top, we'd have to add it specifically to the library.

Reply:
Oh yes indeed now I remember that `Eq` is a derived trait that implements structural equality. Then I shall translate with regular equality in F* or EC :)